File: 2018/Haskell/src/Day06.hs
module Day06 where

import Lib ( freqs )

import qualified Data.Map as Map
import           Data.Map   ( Map )
import Text.ParserCombinators.Parsec
import Data.Semigroup
import Data.List.Extra ( minimumOn )
import Data.Maybe (mapMaybe)

type Pair = (Int, Int)
type Coords = [Pair]

data Bounds =
  Bounds
  { left :: Int
  , top :: Int
  , right :: Int
  , bottom :: Int
  }
  deriving (Show, Eq, Ord)

parseInput :: String -> Coords
parseInput = either (error "Bad parse") id . traverse (parse parseTup "Tuple") . lines

parseTup :: Parser (Int, Int)
parseTup = do
  x <- read <$> many digit
  string ", "
  y <- read <$> many digit
  pure (x, y)

bounds :: Coords -> Bounds
bounds cs = Bounds (getMin left) (getMin top) (getMax right) (getMax bottom)
  where
    (left, top, right, bottom) = foldMap (\(x, y) -> (Min x, Min y, Max x, Max y)) cs

manhatten :: (Int, Int) -> (Int, Int) -> Int
manhatten (x1, y1) (x2, y2) = abs (x2 - x1) + abs (y2 - y1)

closestPoint :: Pair -> Coords -> Maybe Pair
closestPoint (x,y) cs =
  let candidates = [ (p, manhatten (x,y) p) | p <- cs ]
      shortest = minimumOn snd candidates
      count = length $ filter ((== snd shortest) . snd) candidates
   in case count of
        1 -> Just $ fst shortest
        _ -> Nothing

findNearest :: Bounds -> Coords -> Map Pair (Maybe Pair)
findNearest (Bounds l t r b) cs =
  Map.fromList [ ((x, y), pair)
               | x <- [l .. r]
               , y <- [t .. b]
               , let pair = closestPoint (x,y) cs
               ]

removeInfinite :: Bounds -> Map Pair (Maybe Pair) -> Map Pair (Maybe Pair)
removeInfinite (Bounds l t r b) cands =
  let left = zip (repeat l) [t .. b]
      top = zip (repeat t) [l .. r]
      right = zip (repeat r) [t .. b]
      bottom = zip (repeat b) [l .. r]
      edge = mapMaybe (`Map.lookup` cands) $ left ++ right ++ top ++ bottom
   in Map.filter (not . (`elem` edge)) cands

part1 :: Map Pair (Maybe Pair) -> Int
part1 = maximum . freqs . Map.elems

allDists :: (Int, Int) -> [(Int, Int)] -> Int
allDists p cs = sum [ manhatten p c | c <- cs ]

findAllLocations :: Int -> Bounds -> Coords -> Int
findAllLocations maxDist (Bounds l t r b) cs =
  length [ (x,y)
         | x <- [l .. r]
         , y <- [t .. b]
         , allDists (x,y) cs < maxDist
         ]

part2 :: Bounds -> Coords -> Int
part2 = findAllLocations 10000

main :: IO ()
main = do
  input <- parseInput <$> readFile "../data/day06.in"
  let bs = bounds input
      coords = removeInfinite bs . findNearest bs  $ input
  print $ part1 coords
  print $ part2 bs input

-- 4233
-- 45290

File: 2018/Python/day06.py
from string import ascii_lowercase
from string import ascii_uppercase
from itertools import chain

def manhatten_distance(coord, keys, data):
    col, row = coord
    res = data.copy()
    for a, b in keys:
        dist = abs(col - a) + abs(row - b)
        res[(a, b)] = dist
    rv = min(res, key=lambda x: res[x])
    if list(res.values()).count(res[rv]) > 1:
        return False
    return rv

def read_file(filename):
    data = {}
    keys = []
    with open('data/day06.in') as f:
        for line in f:
            col, row = (int(i) for i in line.split(", "))
            data[(col, row)] = 0
            keys.append((col, row))
    return data, keys

def solveA(data, keys):
    coords = dict(zip(data.keys(), chain(ascii_uppercase, ascii_lowercase)))
    for row in range(350):
        for col in range(400):
            coords = manhatten_distance((col, row), keys, data)
            if not coords:
                continue
            if data[coords] == -1:
                continue
            elif row == 0 or col == 0 or row == 349 or col == 399:
                data[coords] = -1
            else:
                data[coords] += 1

    print('Part 1:', max(data.values()))

# Part 2

def manhatten_sum(coord, keys, data):
    col, row = coord
    res = 0
    for a, b in keys:
        res += abs(col - a) + abs(row - b)
    return res

def solveB(keys, data):
    count = 0
    for row in range(350):
        for col in range(400):
            dist = manhatten_sum((col, row), keys, data)
            if dist < 10000:
                count += 1
    print('Part 2:', count)

def main():
    data, keys = read_file('data/day06.in')
    solveA(data, keys)
    solveB(keys, data)


if __name__ == '__main__':
    main()
